Sources/SwiftSyntax/SymbolName.swift

```diff
@@ -0,0 +1,142 @@
+  /// The symbol name is a string representation of the subscript along with its parameter labels.
+  /// For example, a subscript `subscript(index: Int)` will have the symbol name `subscript(_:)`.
+  public var symbolName: String {
+    SwiftSyntax.symbolName(
+      fromBaseName: "subscript",
+      andParameters: parameterClause
+    )
+  }
+}
+
+extension EnumCaseElementSyntax {
+  /// The symbol name of the enum case element.
+  ///
+  /// The symbol name is a string representation of the enum case name along with its associated value labels (if any).
+  /// For example, an enum case `case foo(bar: Int)` will have the symbol name `foo(bar:)`.
+  public var symbolName: String {
+    let caseName = name.text
+
+    guard let associatedValue = parameterClause else {
+      return caseName
+    }
+
+    let argumentNames = associatedValue.parameters
+      .map { parameter in
+        guard let firstName = parameter.firstName else {
+          return "_:"
+        }
+
+        return firstName.text + ":"
+      }
+      .joined()
+
+    return "\(caseName)(\(argumentNames))"
+  }
+}
+
+extension MacroDeclSyntax {
+  /// The symbol name of the macro declaration.
+  ///
+  /// The symbol name is a string representation of the macro name along with its parameter labels.
+  /// For example, a macro `macro greet(name: String)` will have the symbol name `greet(name:)`.
+  public var symbolName: String {
+    SwiftSyntax.symbolName(
+      fromBaseName: name.trimmedDescription,
+      andParameters: signature.parameterClause
+    )
+  }
+}
+
+/// Generates the symbol name by combining the base name and parameter labels.
+///
+/// - Parameters:
+///   - baseName: The base name of the symbol (e.g., function name, initializer, subscript).
+///   - parameters: The function parameter clause containing the parameter labels.
+/// - Returns: The symbol name with the base name and parameter labels combined.
+private func symbolName(
+  fromBaseName baseName: String,
+  andParameters parameters: FunctionParameterClauseSyntax
+) -> String {
+  let argumentNames = parameters.parameters
+    .map { parameter in
+      let argumentLabelText = parameter.argumentName?.text ?? "_"
+      return argumentLabelText + ":"
+    }
+    .joined()
+
+  return "\(baseName)(\(argumentNames))"
+}
+
+extension FunctionParameterSyntax {
+  /// The argument name (label) of the function parameter.
+  ///
+  /// If the parameter has two names (e.g., `external internal: Int`), the first name is considered the argument label.
+  /// If the parameter has only one name and it's not a subscript parameter, it is considered the argument label.
+  fileprivate var argumentName: TokenSyntax? {
+    // If we have two names, the first one is the argument label
+    if secondName != nil {
+      return firstName.asIdentifierToken
+    }
+
+    // If we have only one name, it might be an argument label.
+    if let superParent = parent?.parent?.parent, superParent.is(SubscriptDeclSyntax.self) {
```

Comment:
Instead of walking the tree up again (which might be fragile if for some reasons the `FunctionParameterSyntax` would occur eg. in the subscript body and just happens to be 3 levels down from the `SubscriptDeclSyntax` node – not that this might happen today but could theoretically in the future), I’d just add a parameter `isForSubscript`.

And then I’d probably flip the implementation around and have a `if isForSubscript` at the top level. I think that’s easier to read.

Reply:
Good point; done